scripts/run/update.sh

```diff
@@ -1,78 +1,241 @@
+      
+      echo "Attempting to start..."
+      systemctl start "${servname}"
+
+      if ! systemctl is-active --quiet "${servname}"; then
+        echo "${FAIL} ${RED}Unable to start '${servname}'${RESET}"
+        rollback_changes || echo "${CRITICAL} ${RED}Unable to rollback changes!${RESET}" 
+        return 1
+      fi
+    fi
+    echo "${OK} '${servname}' is active."
   done
+
+  return 0
+}
+
+#######################################
+# Rolls back the most recent changes by resetting the HEAD to the previous
+# commit and then restores the contents of the service files.
+# Globals:
+#   None
+# Arguments:
+#   None
+# Outputs:
+#   Warning message about rolling back changes and the results of updating
+#   services after the rollback.
+# Returns:
+#   Status of update_services function.
+#######################################
+function rollback_changes() {
+  echo "${WARNING} ${YELLOW}Rolling back changes...${RESET}"
+
+  git reset --hard HEAD~1
+  return "$(update_services)"
 }
 
+#######################################
+# Main function to ensure network connectivity, pull latest changes from a
+# specified branch, update service files, and check the system status. If
+# the network is down, it attempts to connect to WiFi.
+# Globals:
+#   OK, GREEN, WARNING, YELLOW, FAIL, RED, RESET - Used for coloring messages.
+# Arguments:
+#   None
+# Outputs:
+#   Various status messages about network connectivity, Git operations, service
+#   file updates, and system status checks.
+#######################################
 function main() {
-  local scripts=("post-reboot/set_user_permissions.sh" "pre-reboot/create_services.sh")
+  if wget -q --spider "https://google.com"; then
+    echo "${OK} ${GREEN}Network connection successful.${RESET}"
+  else
+    echo "${WARNING} ${YELLOW}Network connection failed.${RESET}"
+
+    declare response
+    while true; do
+      read -rp "Do you wish to connect to WiFi? [Y/n] " response
+      case "${response}" in
+        [Yy])
+          break
+          ;;
+        [Nn])
+          return 1
+          ;;
+        *)
+          echo "Invalid input: Answer either 'y' (yes) or 'n' (no)."
+          echo
+          ;;
+      esac
+    done
+
+    declare -r max_attempts=3
+    for ((attempt = 1; attempt <= max_attempts; attempt++)); do
+      echo "Attempt ${attempt} of ${max_attempts}:"
+      if connect_network; then
+        echo "${OK} ${GREEN}Network connection successful.${RESET}"
+        break
+      fi
+      echo -e "Retrying...\n"
+    done
+
+    if ((attempt > max_attempts)); then
+      echo "${FAIL} ${RED}Network connection failed: Max retry attempts reached.${RESET}"
+      return 1
+    fi
+  fi
+  echo
+
+  declare branch;
+  read -rp "Specify target branch (default: main): " branch
+  if [ -z "${branch}" ]; then
+    branch="main"
+  fi
+
+  pull_latest_changes "${branch}" || return 1
+  echo -e "${OK}${GREEN}Pulled all changes from '${branch}'.${RESET}\n"
```

Comment:
Does the -e conflict with the formatting of `OK`, `GREEN`, etc?

Reply:
No. The way we set up our colors allows us to use `-e` option in `echo`. We are not using escape codes to setup our colors.